chore!: update rollup commonjs plugin to v22

haoqunjiang opened the pull request:
### Description

Cherry-picked a commit from @sapphi-red in https://github.com/vitejs/vite/pull/7894


### Additional context


---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [x] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

sapphi-red committed:
fix: exclude ts-node from bundle

haoqunjiang committed:
chore: update commonjs plugin to v22 and make build work

Well, the built bundle doesn't work yet.
Blocked by https://github.com/rollup/plugins/pull/1180

haoqunjiang committed:
build: resolve `import.meta.url` at build time

sodatea commented on packages/vite/LICENSE.md:
```diff
@@ -2197,7 +2197,7 @@ Repository: unjs/mlly
 > MIT License
 > 
-> Copyright (c) 2022 UnJS
+> Copyright (c) 2021
```
I've no idea why this is happening… All I do was rebuilding the project with the updated dependencies.

haoqunjiang committed:
build: use `pathToFileURL` to support Windows paths

patak-dev commented:
IIUC https://github.com/vitejs/vite/pull/7894 was failing in our test cases. Shouldn't we merge this if the vite CI passes with all playgrounds using `legacy.buildRollupPluginCommonjs`? It feels like an unsafe update if not, and if few people are going to use it, better they get what they knew it was working before, no?

sodatea commented:
Ah, thanks for the reminder. I forgot the tests currently aren't testing `legacy.buildRollupPluginCommonjs`… I just thought I'd somehow passed the CI. Turns out it's far from complete.

haoqunjiang committed:
chore: update commonjs plugin to v22.0.1

haoqunjiang committed:
test: add tests for legacy cjs plugin

haoqunjiang committed:
test: add commonjsOptions to the `external` test

…so that it can correctly transpile `require` to `import *`

haoqunjiang committed:
test: fix env var name 😅

haoqunjiang committed:
ci: add cjs test to CI workflow

haoqunjiang committed:
test: should apply the test config before resolving…

haoqunjiang committed:
test: use the correct id of browserExternalId

In production there's no `:fs` suffix
https://github.com/vitejs/vite/blob/32712668f6586734afa592384e325e9a8b528fe9/packages/vite/src/node/plugins/resolve.ts#L328-L330

haoqunjiang committed:
test: skip the esbuild-plugin test when testing the rollup cjs plugin

sodatea commented:
Tests are added and passed!

Huge thanks to @lukastaegert for the quick release of the patch!

patak-dev commented on package.json:
```diff
@@ -17,9 +17,10 @@
     "lint": "eslint --cache packages/*/{src,types,__tests__}/** playground/**/__tests__/**/*.ts scripts/**",
     "typecheck": "tsc -p scripts --noEmit && tsc -p playground --noEmit",
-    "test": "run-s test-unit test-serve test-build",
+    "test": "run-s test-unit test-serve test-build test-build-legacy-cjs",
```
This is awesome @sodatea! Now that we saw that tests are passing, maybe we could remove this from the main CI tests? We could add a project to vite-ecosystem-ci for example to check it, but making the CI 30% slower because of this seems too much to me. What do you think?

sodatea replied:
Yeah, sure, that makes sense.

haoqunjiang committed:
ci: remove cjs plugin test from the main CI tests

We'll move it to the ecosystem-ci instead
